File: bin/migrate_drinks.js
// Firebaseのdrinksコレクションを複製し、subcategories配列を追加するスクリプト
// 実行方法: 
// 1. firebase login
// 2. firebase use --add (プロジェクトを選択)
// 3. node bin/migrate_drinks.js

const admin = require('firebase-admin');
const { execSync } = require('child_process');

// Firebase接続の初期化
function initializeFirebase() {
  console.log('Initializing Firebase connection...');
  
  try {
    // 方法１：サービスアカウントキーを使用
    const serviceAccountPath = '/Users/esumi_yuuki/Desktop/OSAKEL/store_map_app/osakel-app-firebase-adminsdk-fbsvc-05d7fd9f22.json';
    const fs = require('fs');
    if (fs.existsSync(serviceAccountPath)) {
      console.log('Using service account key for authentication...');
      const serviceAccount = require(serviceAccountPath);
      admin.initializeApp({
        credential: admin.credential.cert(serviceAccount)
      });
    } 
    // 方法２：上記が失敗した場合は、プロジェクトIDを直接指定
    else {
      console.log('Service account key not found. Using project ID directly.');
      // Firebase CLIからプロジェクトIDを取得
      const projectId = execSync('firebase use --json')
        .toString()
        .trim();
        
      const projectInfo = JSON.parse(projectId);
      console.log(`Using Firebase project ID: ${projectInfo.result}`);
      
      // 認証情報を直接指定
      admin.initializeApp({
        projectId: projectInfo.result,
        // GOOGLE_APPLICATION_CREDENTIALS環境変数に依存します
        // ここで、Google Cloudの認証情報を指定することも可能です
      });
    }
    
    console.log('Firebase initialized successfully');
    return admin.firestore();
  } catch (error) {
    console.error('Error initializing Firebase:', error);
    console.error('Please provide a valid service account key file named firebase-credentials.json');
    process.exit(1);
  }
}

// データ移行の実行
async function migrateAndDuplicateDrinks() {
  const db = initializeFirebase();
  
  try {
    console.log('Starting drinks migration...');
    
    // カテゴリとサブカテゴリの情報を取得
    console.log('Fetching categories...');
    const categoriesSnapshot = await db.collection('categories').get();
    const categoryMap = {};
    
    categoriesSnapshot.forEach(doc => {
      const data = doc.data();
      categoryMap[doc.id] = data.subcategories || [];
      console.log(`Category ${doc.id} has ${(data.subcategories || []).length} subcategories`);
    });
    
    // 既存のドリンクを取得
    console.log('Fetching existing drinks...');
    const drinksSnapshot = await db.collection('drinks').get();
    console.log(`Found ${drinksSnapshot.size} drinks to migrate.`);
    
    let batch = db.batch();
    let count = 0;
    let batchCount = 0;
    
    for (const doc of drinksSnapshot.docs) {
      const drink = doc.data();
      const categoryId = drink.categoryId || drink.category || '';
      
      const availableSubcategories = categoryMap[categoryId] || [];
      let selectedSubcategories = [];
      
      if (availableSubcategories.length >= 2) {
        // ランダムに2つ選択
        const shuffled = [...availableSubcategories].sort(() => 0.5 - Math.random());
        selectedSubcategories = shuffled.slice(0, 2).map(sub => {
          return typeof sub === 'object' ? (sub.id || '') : sub.toString();
        });
      } else if (availableSubcategories.length > 0) {
        selectedSubcategories = availableSubcategories.map(sub => {
          return typeof sub === 'object' ? (sub.id || '') : sub.toString();
        });
      }
      
      // 既存のsubcategoryIdがあれば配列に追加（重複しないように）
      if (drink.subcategoryId && !selectedSubcategories.includes(drink.subcategoryId)) {
        selectedSubcategories.push(drink.subcategoryId.toString());
      }
      
      // 新しいドキュメントIDを生成
      const newDocId = `${doc.id}_duplicated`;
      const newDocRef = db.collection('drinks').doc(newDocId);
      
      // 複製したドキュメントを作成
      const newDrink = { ...drink, subcategories: selectedSubcategories };
      
      // バッチに追加
      batch.set(newDocRef, newDrink);
      
      count++;
      
      // Firestoreのバッチ制限（500）に達したらコミット
      if (count % 400 === 0) {
        batchCount++;
        console.log(`Committing batch ${batchCount} with ${count} documents...`);
        await batch.commit();
        batch = db.batch(); // 新しいバッチを作成
      }
    }
    
    // 残りをコミット
    if (count % 400 !== 0) {
      batchCount++;
      console.log(`Committing final batch ${batchCount}...`);
      await batch.commit();
    }
    
    console.log(`Successfully migrated and duplicated ${count} drinks in ${batchCount} batches.`);
    process.exit(0);
  } catch (error) {
    console.error('Error during migration:', error);
    process.exit(1);
  }
}

// スクリプトの実行
migrateAndDuplicateDrinks();

File: bin/migrate_drinks.dart
// Firebaseのdrinksコレクションを複製し、subcategories配列を追加するスクリプト
// 実行方法: flutter run --release -d <device_id> -t bin/migrate_drinks.dart

import 'dart:async';
import 'package:firebase_core/firebase_core.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import '../lib/firebase_options.dart';

void main() async {
  // Flutter初期化（最初に実行する必要がある）
  WidgetsFlutterBinding.ensureInitialized();
  
  runApp(const MigrationApp());
}

class MigrationApp extends StatelessWidget {
  const MigrationApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MigrationScreen(),
    );
  }
}

class MigrationScreen extends StatefulWidget {
  @override
  _MigrationScreenState createState() => _MigrationScreenState();
}

class _MigrationScreenState extends State<MigrationScreen> {
  String _status = 'Initializing...';
  bool _isRunning = false;
  String _result = '';
  
  @override
  void initState() {
    super.initState();
    _initFirebase();
  }
  
  Future<void> _initFirebase() async {
    try {
      setState(() {
        _status = 'Initializing Firebase...';
      });
      
      await Firebase.initializeApp(
        options: DefaultFirebaseOptions.currentPlatform,
      );
      
      setState(() {
        _status = 'Firebase initialized. Ready to migrate.';
      });
    } catch (e) {
      setState(() {
        _status = 'Firebase initialization error: $e';
      });
    }
  }
  
  Future<void> _startMigration() async {
    if (_isRunning) return;
    
    setState(() {
      _isRunning = true;
      _status = 'Migration in progress...';
    });
    
    try {
      final result = await migrateAndDuplicateDrinksForTesting();
      
      setState(() {
        _result = result;
        _status = 'Migration completed.';
        _isRunning = false;
      });
    } catch (e) {
      setState(() {
        _status = 'Error during migration: $e';
        _isRunning = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Drinks Migration Tool'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Text(
              'Status: $_status',
              style: TextStyle(fontSize: 16),
            ),
            const SizedBox(height: 20),
            if (_result.isNotEmpty)
              Padding(
                padding: const EdgeInsets.all(16.0),
                child: Text(
                  'Result: $_result',
                  style: TextStyle(fontSize: 14),
                ),
              ),
            const SizedBox(height: 20),
            ElevatedButton(
              onPressed: _isRunning ? null : _startMigration,
              child: Text(_isRunning ? 'Running...' : 'Start Migration'),
            ),
          ],
        ),
      ),
    );
  }
}

/// 既存のdrinksコレクションを複製し、subcategories配列を追加する
Future<String> migrateAndDuplicateDrinksForTesting() async {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  
  try {
    // カテゴリとサブカテゴリの情報を取得
    debugPrint('Fetching categories...');
    final categoriesSnapshot = await _firestore.collection('categories').get();
    final categoryMap = <String, List<dynamic>>{};
    
    for (var doc in categoriesSnapshot.docs) {
      final data = doc.data();
      categoryMap[doc.id] = data['subcategories'] ?? [];
      debugPrint('Category ${doc.id} has ${(data['subcategories'] ?? []).length} subcategories');
    }
    
    // 既存のドリンクを取得
    debugPrint('Fetching existing drinks...');
    final drinksSnapshot = await _firestore.collection('drinks').get();
    debugPrint('Found ${drinksSnapshot.docs.length} drinks to migrate.');
    
    var batch = _firestore.batch();
    int count = 0;
    
    for (var doc in drinksSnapshot.docs) {
      final drink = doc.data();
      final categoryId = drink['categoryId'] ?? drink['category'] ?? '';
      
      final availableSubcategories = categoryMap[categoryId] ?? [];
      List<String> selectedSubcategories = [];
      
      if (availableSubcategories.length >= 2) {
        // ランダムに2つ選択
        final shuffled = List.from(availableSubcategories)..shuffle();
        selectedSubcategories = shuffled.take(2).map<String>((sub) {
          return sub is Map ? (sub['id'] ?? '') : sub.toString();
        }).toList();
      } else if (availableSubcategories.isNotEmpty) {
        selectedSubcategories = availableSubcategories.map<String>((sub) {
          return sub is Map ? (sub['id'] ?? '') : sub.toString();
        }).toList();
      }
      
      // 既存のsubcategoryIdがあれば配列に追加（重複しないように）
      if (drink['subcategoryId'] != null && 
          !selectedSubcategories.contains(drink['subcategoryId'])) {
        selectedSubcategories.add(drink['subcategoryId'].toString());
      }
      
      // 新しいドキュメントIDを生成
      final newDocId = '${doc.id}_duplicated';
      final newDocRef = _firestore.collection('drinks').doc(newDocId);
      
      // 複製したドキュメントを作成
      final Map<String, dynamic> newDrink = Map.from(drink);
      newDrink['subcategories'] = selectedSubcategories;
      
      // バッチに追加
      batch.set(newDocRef, newDrink);
      
      count++;
      
      // Firestoreのバッチ制限（500）に達したらコミット
      if (count % 400 == 0) {
        debugPrint('Committing batch of $count documents...');
        await batch.commit();
        batch = _firestore.batch(); // 新しいバッチを作成
      }
    }
    
    // 残りをコミット
    if (count % 400 != 0) {
      debugPrint('Committing final batch...');
      await batch.commit();
    }
    
    return 'Successfully migrated and duplicated $count drinks.';
  } catch (error) {
    debugPrint('Error during migration: $error');
    return 'Migration failed: ${error.toString()}';
  }
}
